File: core/src/main/java/heroes/journey/entities/items/ConsumableItem.java
package heroes.journey.entities.items;

import static heroes.journey.registries.Registries.ItemManager;

import java.util.function.Consumer;

import heroes.journey.entities.actions.inputs.ActionInput;
import heroes.journey.entities.tagging.Attributes;

public class ConsumableItem extends Item {
  protected final Consumer<ActionInput> onConsume;

  public ConsumableItem(
      String id,
      String name,
      ItemSubType subType,
      int weight,
      int value,
      Attributes attributes,
      Consumer<ActionInput> onConsume) {
    super(id, name, subType, weight, value, attributes);
    this.onConsume = onConsume;
  }

  public void consume(ActionInput input) {
    onConsume.accept(input);
  }

  public ConsumableItem register() {
    ItemManager.register(this);
    return this;
  }
}

File: core/src/main/java/heroes/journey/entities/items/ConsumableItem.kt
package heroes.journey.entities.items

import heroes.journey.entities.actions.ActionContext
import heroes.journey.entities.tagging.Attributes
import heroes.journey.registries.Registries.ItemManager
import java.util.function.Consumer

class ConsumableItem(
    id: String,
    subType: ItemSubType,
    weight: Int,
    attributes: Attributes = Attributes(),
    protected val onConsume: Consumer<ActionContext>
) : Item(id, subType, weight, attributes) {

    fun consume(input: ActionContext) {
        onConsume.accept(input)
    }

    override fun register(): ConsumableItem = this.apply { ItemManager.register(this) }
}
